### ISSUE TITLE
README Device Address Service Example Not Working

### ISSUE THREAD

**Issue by Rabadash8820** [alexa/alexa-skills-kit-sdk-for-nodejs#307]
In the main README for this repo, under the [Directive Service](https://github.com/alexa/alexa-skills-kit-sdk-for-nodejs#directive-service) heading, there is some example code showing how to use directives in an intent handler, including calls to `Promise.all().then()`.  Unfortunately, I have tried copying this code verbatim, and am unable to get a response.  My Lambda function just times out after the default three seconds without any other error messages.  Does this example need to be updated or am I missing something?

To be clear, this is the code that I'm using:
```js
const Alexa = require('alexa-sdk');

const handlers = {
    'SearchIntent' : function() {
        const requestId = this.event.request.requestId;
        const token = this.event.context.System.apiAccessToken;
        const endpoint = this.event.context.System.apiEndpoint;
        const ds = new Alexa.services.DirectiveService();

        const directive = new Alexa.directives.VoicePlayerSpeakDirective(requestId, "Please wait...");
        const progressiveResponse = ds.enqueue(directive, endpoint, token)
        .catch((err) => {
            // catch API errors so skill processing an continue
        });

        Promise.all([progressiveResponse])
        .then(() => {
            this.response.speak('I found the following results');
            this.emit(':responseReady');
        });
    }
};
```

I think the following issues are related, but I wasn't able to find a solution in either of them:
#44
#29

**Comment by allthepies** [comment]
Hmmm, the progressive response works for me.

A few things to try:

1. Put some consol.log() debug in the catch() handler for the enqueue request.
2. Increase your lambda timeout to 10s

I didn't wait for the progressiveResponse promise to resolve in my case, I issued the enqueue() and then started performing the heavy lifting code which I was notifying the user about.

**Comment by Rabadash8820** [comment]
1. I added some `console.log()` statements to the `catch()` but their messages didn't show up.
2. Increasing the timeout didn't help.

I'm concerned that the issue is not actually with the use of directives but with the use of Promises.  I've been working through a number of online examples involving Promises (both native and [bluebird](http://bluebirdjs.com/docs/getting-started.html)) and all of them are giving me timeout errors

**Comment by allthepies** [comment]
Just tried the following in my skill:

```
function updateUser(event, updateText) {
  const requestId = event.request.requestId;
  const token = event.context.System.apiAccessToken;
  const endpoint = event.context.System.apiEndpoint;

  const ds = new Alexa.services.DirectiveService();

  const directive = new Alexa.directives.VoicePlayerSpeakDirective(requestId, updateText);
  const prog = ds.enqueue(directive, endpoint, token)
    .catch((err) => {
      // catch API errors so skill processing an continue
      console.log('Error in updateUser: ' + JSON.stringify(err));

    });

  Promise.all([prog])
    .then(() => {
      console.log('PROMISE RESOLVED');
    });
}
```

And get PROMISE RESOLVED output + my progressive response.

What happens if you take the response.speak and emit out of the Promise.all block and put them inline just afterwards ?

**Comment by Rabadash8820** [comment]
Still getting a timeout, no additional errors 🙁  I wasn't quite sure what you meant by "put them inline just afterwards" though, so my current intent handler is shown below.  Is that what you meant?

```js
"SearchIntent": function() {
        updateUser(this.event, "Please wait...");

        this.response.speak("I found the following results");
        this.emit(":responseReady");
        
        function updateUser(event, updateText) {
            const requestId = event.request.requestId;
            const token = event.context.System.apiAccessToken;
            const endpoint = event.context.System.apiEndpoint;

            const ds = new Alexa.services.DirectiveService();

            const directive = new Alexa.directives.VoicePlayerSpeakDirective(requestId, updateText);
            const prog = ds.enqueue(directive, endpoint, token)
              .catch((err) => {
                  // catch API errors so skill processing an continue
                  console.log("Error in updateUser: " + JSON.stringify(err));
              });

            Promise.all([prog])
              .then(() => {
                  console.log("PROMISE RESOLVED");
              });
        }
    }
```

Thanks again for your help.

**Comment by allthepies** [comment]
Ah, no.  Don't include my function in your function, go back to what you had.

Are you sure the SearchIntent is being entered ?  Add a console.log() output at various places within your function.  i.e. on entry, after setting "ds", after setting "directive" and after invoking "ds,enqueue".

Do all of those console.log outputs appear ?

**Comment by Rabadash8820** [comment]
`SearchIntent` is being entered; I placed `console.log()` statements where you suggested and all of them appear in the log.  However, I'm getting the following behaviour:

I tried pulling out the `speak`/`emit` lines with a preceding `console.log()`, like you suggested above:

```
Promise.all([progressiveResponse]);

console.log("Emitting response...");
this.response.speak("I found the following results");
this.emit(":responseReady");
```

which showed all `console.log()` statements:

```
START RequestId: 1834f5a1-1c08-11e8-b9dd-e728ea1309c3 Version: $LATEST
2018-02-27T21:49:33.560Z	1834f5a1-1c08-11e8-b9dd-e728ea1309c3	Entering...
2018-02-27T21:49:33.579Z	1834f5a1-1c08-11e8-b9dd-e728ea1309c3	ds defined...
2018-02-27T21:49:33.980Z	1834f5a1-1c08-11e8-b9dd-e728ea1309c3	ds enqueued...
2018-02-27T21:49:33.999Z	1834f5a1-1c08-11e8-b9dd-e728ea1309c3	Emitting response...
END RequestId: 1834f5a1-1c08-11e8-b9dd-e728ea1309c3
REPORT RequestId: 1834f5a1-1c08-11e8-b9dd-e728ea1309c3	Duration: 3003.27 ms	Billed Duration: 3000 ms 	Memory Size: 128 MB	Max Memory Used: 47 MB	
```

However, using the code from my first comment and adding `console.log("Promise then...")` line just inside the `Promise.all().then()` yielded the following output:

```
START RequestId: bf9ebc93-1c07-11e8-a491-5b19337b9cf7 Version: $LATEST
2018-02-27T21:47:04.963Z	bf9ebc93-1c07-11e8-a491-5b19337b9cf7	Entering...
2018-02-27T21:47:05.019Z	bf9ebc93-1c07-11e8-a491-5b19337b9cf7	ds defined...
2018-02-27T21:47:05.719Z	bf9ebc93-1c07-11e8-a491-5b19337b9cf7	ds enqueued...
END RequestId: bf9ebc93-1c07-11e8-a491-5b19337b9cf7
REPORT RequestId: bf9ebc93-1c07-11e8-a491-5b19337b9cf7	Duration: 3003.24 ms	Billed Duration: 3000 ms 	Memory Size: 128 MB	Max Memory Used: 47 MB	
```

So it looks like the `Promise.then()` is not being called.  I found [this article](https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/using-promises.html) that goes into some detail about using Promises in AWS lambda functions.  From what I can tell, my environment should support Promises, but maybe I forgot to configure something?

**Comment by allthepies** [comment]
I use promises all over the place in my Alexa skills (all are AWS lambda based) so they definitely work.

When I use the progressive response I don't wait for it to complete so you could just omit the .then() from the Promise.all.  Mind you, with just one entry in the promise array then you might as well do without the Promise.all() statement too.

**Comment by Rabadash8820** [comment]
So where do you put the final `speak`/`emit`?  I just tried deleting the `Promise.all()` block.  Both putting the emit statements at the end of the handler, and putting them inside `enqueue().then()`, are still timing out.

**Comment by Rabadash8820** [comment]
So the only reason I was following the Directive Service example in the first place was because I was trying to use Promises in the actual skill that I was working on, and I just wanted to find *some* example of Promises working in an Alexa skill.  Everything I had tried kept timing out.  However, it looks like these timeouts were occurring not because of any bug in the code, but because of my Lambda function's network configuration.  It was trying to execute the function in a VPC, which breaks public Internet access, which I think that maybe the Directive Service requires.  [This article](https://aws.amazon.com/premiumsupport/knowledge-center/internet-access-lambda-function/) was very helpful in fixing my infrastructure.

I didn't test out the Directive Service example after fixing my infrastructure; however, if @allthepies got it to work before then I'm sure that there's no issue with it.  Closing this issue now...
